### PR TITLE
Add Subdomain Enumeration Methods

### PR DESCRIPTION
**Pull request by kshitijk4poor**
This PR introduces a new module, subdomain_enumeration.py, designed to gather subdomains for a given domain using various tools. The aim of this module is to enhance the capability of the system to gather subdomains efficiently and reliably, leveraging multiple tools and robust error handling.

- [x] Retry Mechanism
- [x] Uniform Tool Integration Methods
- [x] Redis Caching pipeline
- [x] Verbose Logging
- [x] Domain Validation
- [x] Testcases
- [x] Code refactor for better Maintenance, Readability and Optimization

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -4,9 +4,9 @@ x-artemis-build-or-image:
   # Replace with the following to use local Artemis source code to build the container,
   # not one downloaded from Docker Hub:
   #
-  # build:
-  #    context: .
-  #    dockerfile: docker/Dockerfile
+  build:
```
Comment (kazet) [inline docker-compose.yaml:7]: don't commit that change ;)

File: docker-compose.yaml [hunk 2]
```diff
@@ -94,7 +94,7 @@ services:
     <<: *artemis-build-or-image
     depends_on: [karton-system, karton-logger]
     env_file: .env
-    ports: ["5000:5000"]
+    ports: ["7500:5000"]
```
Comment (kazet) [inline docker-compose.yaml:97]: don't comit that change ;)

File: docker-compose.yaml [hunk 3]
```diff
@@ -375,6 +375,14 @@ services:
     restart: always
     volumes: ["./docker/karton.ini:/etc/karton/karton.ini", "${DOCKER_COMPOSE_ADDITIONAL_SHARED_DIRECTORY:-./shared}:/shared/"]
 
+  karton-subdomainenumeration:
```
Comment (kazet) [inline docker-compose.yaml:378]: karton-subdomain_enumeration IMO

File: docker/Dockerfile [hunk 4]
```diff
@@ -26,4 +28,4 @@ COPY "artemis/" "artemis/"
 COPY "static/" "static/"
 COPY "templates/" "templates/"
 
-CMD ["uvicorn", "artemis.main:app", "--host", "0.0.0.0", "--port", "5000"]
+CMD ["uvicorn", "artemis.main:app", "--host", "0.0.0.0", "--port", "5000"]
```
Comment (kazet) [inline docker/Dockerfile:31]: no need to change the final newline ;)

File: karton-logs/2024-06-04.log.gz
Comment (kazet) [inline karton-logs/2024-06-04.log.gz:1]: don't commit logs from your machine

File: scripts/start [hunk 5]
```diff
@@ -40,7 +40,7 @@ fi
     $ADDITIONAL_OPTIONS
 
 if [ $? -eq 0 ]; then
-    ./docker/wait-for-it.sh 127.0.0.1:5000 -t 30 -q
+    ./docker/wait-for-it.sh 127.0.0.1:7500 -t 30 -q
```
Comment (kazet) [inline scripts/start:43]: don't change it on main

File: test/modules/test_subdomainenumeration.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+class TestData(NamedTuple):
+    domain: str
+    expected_subdomain: str
+
+
+class SubdomainEnumerationScannerTest(ArtemisModuleTestCase):
+    # The reason for ignoring mypy error is https://github.com/CERT-Polska/karton/issues/201
+    karton_class = subdomainenumeration 
+
+    def test_simple(self) -> None:
```
Comment (kazet) [inline test/modules/test_subdomainenumeration.py:19]: this end-to-end test case makes sense but i'd insist on testing each of get_subdomains_from_... methods

File: artemis/modules/subdomainenumeration.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+from artemis.binds import TaskStatus, TaskType
+from artemis.config import Config
+from artemis.domains import is_subdomain
+from artemis.module_base import ArtemisBase
+from artemis.utils import check_output_log_on_error
+
+DOMAIN_REGEX = r"([a-z0-9\-]+\.)+[a-z0-9\-]+"
+
+class subdomainenumeration(ArtemisBase):
```
Comment (kazet) [inline artemis/modules/subdomainenumeration.py:21]: this class name is not consistent with Puthon best practices

File: artemis/modules/subdomainenumeration.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    Consumes `type: domain` to gather subdomains and produces `type: domain`.
+    """
+
+    identity = "subdomainenumeration"
```
Comment (kazet) [inline artemis/modules/subdomainenumeration.py:26]: imo subdomain_enumeration, easier to read

File: artemis/modules/subdomainenumeration.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    filters = [
+        {"type": TaskType.DOMAIN.value},
+    ]
+    lock_target = False
+
+    def get_subdomains_from_rapid(self, current_task: Task) -> Set[str]:
```
Comment (kazet) [inline artemis/modules/subdomainenumeration.py:32]: use rapiddns - the actual name of the service

File: artemis/modules/subdomainenumeration.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                    td_tag = th_tag.find_next_sibling('td')
+                    if td_tag is not None:
+                        url = td_tag.get_text()
+                        subdomains.add(url)
+        except requests.exceptions.RequestException:
+            self.log.exception("Unable to obtain information from rapiddns.io for domain %s", domain)
+        return subdomains
+
+    def get_subdomains_from_jldc(self, current_task: Task) -> Set[str]:
+        domain = current_task.get_payload("domain")
```
Comment (kazet) [inline artemis/modules/subdomainenumeration.py:50]: `domain = current_task.get_payload("domain")` can be even extracted outside of all get_* methods

File: artemis/modules/subdomainenumeration.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        return subdomains
+
+
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        subdomains = (
+            self.get_subdomains_from_subfinder(current_task) |
+            self.get_subdomains_from_amass(current_task) |
+            self.get_subdomains_from_rapid(current_task) |
+            self.get_subdomains_from_jldc(current_task)
```
Comment (kazet) [inline artemis/modules/subdomainenumeration.py:106]: and what about crtsh and gau?

File: artemis/modules/subdomainenumeration.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        )
+
+        for subdomain in subdomains:
+            if is_subdomain(subdomain, domain):
+                task = Task(
```
Comment (kazet) [inline artemis/modules/subdomainenumeration.py:111]: this way subdomain enumeration will run on all subdomains returned from this module

File: artemis/modules/subdomainenumeration.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            subdomains.update(result.decode().splitlines())
+        except Exception:
+            self.log.exception("Unable to obtain information from amass for domain %s", domain)
+        return subdomains
+
+
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        subdomains = (
+            self.get_subdomains_from_subfinder(current_task) |
```
Comment (kazet) [inline artemis/modules/subdomainenumeration.py:103]: crtsh and gau have retry logic, I'd add it as well

File: artemis/modules/subdomainenumeration.py (hunk 12 above)
Comment (kshitijk4poor) [reply artemis/modules/subdomainenumeration.py]: hi it's intentional because sometimes the testing subdomains have another layer of subdomains.

for eg. 
test1.testing.example.com
test2.testing.example.com

Comment (kazet) [reply artemis/modules/subdomainenumeration.py]: IMO it's better to do this in one pass if possible as this may be a lot faster

File: test/modules/test_subdomain_enumeration.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+from artemis.binds import TaskType
+from artemis.modules.subdomain_enumeration import subdomain_enumeration
+
+
+class TestData(NamedTuple):
+    domain: str
+    expected_subdomain: str
+
+
+class subdomain_enumerationScannerTest(ArtemisModuleTestCase):
```
Comment (kazet) [inline test/modules/test_subdomain_enumeration.py:15]: SubdomainEnumeration

File: docker-compose.yaml [hunk 15]
```diff
@@ -375,6 +361,14 @@ services:
     restart: always
     volumes: ["./docker/karton.ini:/etc/karton/karton.ini", "${DOCKER_COMPOSE_ADDITIONAL_SHARED_DIRECTORY:-./shared}:/shared/"]
 
+  karton-SubdomainEnumeration:
```
Comment (kazet) [inline docker-compose.yaml:364]: here karton-subdomain_enumeration, consistent with the rest of the list

btw, kartons should be alphavetically sorted

File: test/modules/test_subdomain_enumeration.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+            found = False
+            for item in results:
+                if item.payload["domain"] == entry.expected_subdomain:
+                    found = True
+            self.assertTrue(found)
+
+    def test_get_subdomains_from_subfinder(self) -> None:
+        subdomain_enum = SubdomainEnumeration()
+        result = subdomain_enum.et_subdomains_from_subfinder("cert.pl")
```
Comment (kazet) [inline test/modules/test_subdomain_enumeration.py:39]: this cannot pass, use .get_ not .et ;)

File: artemis/modules/SubdomainEnumeration [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        subdomains = (
+            self.get_subdomains_from_subfinder(domain) |
+            self.get_subdomains_from_amass(domain)
+        )
+
+        if self.redis.get(f"crtsh-done-{domain}"):
+            self.log.info(
+                "Crtsh has already returned %s - and as it's a recursive query, no further query will be performed.",
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration:78]: where do you set the key? and why is it named crtsh-done ?

File: artemis/config.py [hunk 18]
```diff
@@ -234,13 +234,6 @@ class Miscellaneous:
             "Artemis modules that are disabled by default (but may easily be enabled in the UI)",
         ] = get_config("MODULES_DISABLED_BY_DEFAULT", default="example,humble", cast=decouple.Csv(str, delimiter=","))
 
-        SUBDOMAIN_ENUMERATION_TTL_DAYS: Annotated[
```
Comment (kazet) [inline artemis/config.py:237]: why is this removed? I think you'll need that

File: artemis/modules/SubdomainEnumeration.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    Consumes `type: domain` to gather subdomains and produces `type: domain`.
+    """
+
+    identity = "SubdomainEnumeration"
+    filters = [
+        {"type": TaskType.DOMAIN.value},
+    ]
+    lock_target = False
+
+    def get_subdomains_with_retry(self, func, domain: str, retries: int = 20) -> Set[str]:
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:29]: add type for func (hint: typing.Callable)

File: artemis/config.py [hunk 20]
```diff
@@ -233,7 +233,7 @@ class Miscellaneous:
             List[str],
             "Artemis modules that are disabled by default (but may easily be enabled in the UI)",
         ] = get_config("MODULES_DISABLED_BY_DEFAULT", default="example,humble", cast=decouple.Csv(str, delimiter=","))
-
+
```
Comment (kazet) [inline artemis/config.py:236]: space in empty line

File: artemis/modules/SubdomainEnumeration.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+        for retry in range(retries):
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:30]: sleep after failed retry so that we don't overload the tools

File: artemis/modules/SubdomainEnumeration.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        subdomains = (
+            self.get_subdomains_with_retry(self.get_subdomains_from_subfinder, domain) |
+            self.get_subdomains_with_retry(self.get_subdomains_from_amass, domain)
+        )
+
+        if self.redis.get(f"SubdomainEnumeration-done-{domain}"):
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:85]: where do you set that key?

File: artemis/modules/SubdomainEnumeration.py [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+            if "panic: runtime error: invalid memory address or nil pointer dereference" in str(e):
+                self.log.error(f"Amass encountered a runtime error for domain {domain}. Skipping subdomain enumeration.")
+            else:
+                self.log.exception("Unable to obtain information from amass for domain %s", domain)
+        return subdomains
+
+
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        subdomains = (
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:80]: you don't create new tasks for subdomains

File: scripts/start (hunk 5 above)
Comment (kazet) [reply scripts/start:43]: this comment hasn't yet been done

File: docker/Dockerfile (hunk 4 above)
Comment (kazet) [reply docker/Dockerfile:30]: this comment hasn't yet been done

File: artemis/modules/SubdomainEnumeration.py [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+    ]
+    lock_target = False
+
+    def get_subdomains_with_retry(self, func: Callable[[str], Set[str]], domain: str, retries: int = 20) -> Set[str]:
+        for retry in range(retries):
+            try:
+                return func(domain)
+            except Exception:
+                self.log.exception("Retry %d/%d for %s", retry + 1, retries, func.__name__)
+                time.sleep(300)
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:35]: make this funtion parameter

File: artemis/modules/SubdomainEnumeration.py [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+            )
+            self.db.save_task_result(task=current_task, status=TaskStatus.OK)
+            return
+ 
+        subdomains = (
+            self.get_subdomains_with_retry(self.get_subdomains_from_subfinder, domain) |
+            self.get_subdomains_with_retry(self.get_subdomains_from_amass, domain)
+        )
+        
+        self.redis.setex(
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:96]: this one is not yet correct

File: artemis/modules/SubdomainEnumeration.py [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+            f"SubdomainEnumeration-done-{encoded_domain}", Config.Miscellaneous.SUBDOMAIN_ENUMERATION_TTL_DAYS * 24 * 60 * 60, 1
+            )
+ 
+        for subdomain in subdomains:
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:100]: check whether subdomain is indeed a subdomain

File: artemis/modules/crtsh.py [hunk 27]
```diff
@@ -1,137 +0,0 @@
...
-                if retry_id < Config.Modules.Crtsh.CRTSH_NUM_RETRIES - 1:
-                    time.sleep(Config.Modules.Crtsh.CRTSH_SLEEP_ON_RETRY_SECONDS)
-                else:
-                    raise UnableToObtainSubdomainsException()
-            else:
-                break
-
-        assert ct_domains is not None
-        for entry in ct_domains:
-            if not is_subdomain(entry, domain):
```
Comment (kazet) [inline artemis/modules/crtsh.py:114]: this

File: artemis/modules/SubdomainEnumeration.py (hunk 26 above)
Comment (kazet) [reply artemis/modules/SubdomainEnumeration.py:100]: copy logic from crtsh

Comment (kazet) [reply artemis/modules/SubdomainEnumeration.py:100]: copy logic from crtsh

File: artemis/modules/SubdomainEnumeration.py [hunk 28]
```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python3
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:1]: the module is not named according to Python module naming convention. for modules it's module_name

File: artemis/modules/SubdomainEnumeration.py [hunk 29]
```diff
@@ -0,0 +1,122 @@
...
+        valid_subdomains = set()
+        for subdomain in subdomains:
+            if not is_subdomain(subdomain, domain):
+                self.log.info("Non-subdomain returned: %s from %s", subdomain, domain)
+                continue
+
+            valid_subdomains.add(subdomain)
+
+        self.redis.setex(
+            f"SubdomainEnumeration-done-{encoded_domain}", Config.Miscellaneous.SUBDOMAIN_ENUMERATION_TTL_DAYS * 24 * 60 * 60, 1
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:106]: still not correct, you need to set it for each subdomain

File: artemis/modules/SubdomainEnumeration.py [hunk 30]
```diff
@@ -0,0 +1,122 @@
...
+    filters = [
+        {"type": TaskType.DOMAIN.value},
+    ]
+    lock_target = False
+
+    def get_subdomains_with_retry(self, func: Callable[[str], Set[str]], domain: str, retries: int = 20, sleep_time: int = 300) -> Set[str]:
+        for retry in range(retries):
+            try:
+                return func(domain)
+            except Exception:
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:34]: hmm, this can't possibly work. you catch exceptions but each of get_subdomains_from won't raise an exceptino as its internals are wrapped inside try...ecxcept

File: artemis/modules/SubdomainEnumeration.py [hunk 31]
```diff
@@ -0,0 +1,122 @@
...
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        encoded_domain = domain.encode("idna")  # using this so that special characters do not break the redis db.
+        
+        subdomains = (
+            self.get_subdomains_with_retry(self.get_subdomains_from_subfinder, domain) |
+            self.get_subdomains_with_retry(self.get_subdomains_from_amass, domain)
+        )
+
+        if self.redis.get(f"SubdomainEnumeration-done-{encoded_domain}"):
```
Comment (kazet) [inline artemis/modules/SubdomainEnumeration.py:89]: imo a good naming convention for redis keys would be lowercase but that's not very important

File: artemis/modules/subdomain_enumeration.py [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+
+            self.log.info("Retrying for domain %s", domain)
+            if retry_id < retries - 1:
+                time.sleep(sleep_time)
+            else:
+                raise UnableToObtainSubdomainsException(f"Unable to obtain subdomains for {domain} after {retries} retries")
+        
+        return set()
+
+    def get_subdomains_from_subfinder(self, domain: str) -> Optional[Set[str]]:
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:45]: I insist on adding gau, as the list you provided from cert.pl skipped some domains that gau gives us (e.g. nomoreransom.cert.pl)

File: artemis/modules/subdomain_enumeration.py [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+                    "-d", 
+                    domain, 
+                    "-silent",
+                    "-all",
+                    "-recursive"
+                    ],
+                    self.log
+            )
+            subdomains.update(result.decode().splitlines())
+            print(f"Subdomains found from subfinder: {subdomains}")
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:60]: print -> self.log

btw, this log can be in line ~95

File: artemis/modules/subdomain_enumeration.py [hunk 34]
```diff
@@ -0,0 +1,145 @@
...
+        print(f"Running subdomain enumeration for domain: {domain}")
+
+        try:
+            subdomains = (
+                self.get_subdomains_with_retry(self.get_subdomains_from_subfinder, domain) |
+                self.get_subdomains_with_retry(self.get_subdomains_from_amass, domain)
+            )
+            self.log.debug(f"Subdomains found: {subdomains}")
+            print(f"Subdomains found: {subdomains}")
+        except UnableToObtainSubdomainsException as e:
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:100]: this way if subfinder breaks you won't run amass - maybe try...except should be done per-datasource?

File: artemis/modules/subdomain_enumeration.py [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+            self.log.error(f"Failed to obtain subdomains for domain {domain}: {e}")
+            print(f"Failed to obtain subdomains for domain {domain}: {e}")
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:102]: print -> self.log

File: artemis/modules/subdomain_enumeration.py [hunk 36]
```diff
@@ -0,0 +1,145 @@
...
+
+            valid_subdomains.add(subdomain)
+
+            encoded_subdomain = subdomain.encode("idna").decode("utf-8") # using this so that special characters do not break the redis db.
+            if self.redis.get(f"subdomainenumeration-done-{encoded_subdomain}"):
+                self.log.info(
+                    "SubdomainEnumeration has already returned %s - and as it's a recursive query, no further query will be performed.",
+                    domain,
+                )
+                print(f"SubdomainEnumeration has already returned {subdomain} - skipping further queries.")
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:121]: print -> self.log

File: artemis/modules/subdomain_enumeration.py [hunk 37]
```diff
@@ -0,0 +1,145 @@
...
+                self.db.save_task_result(task=current_task, status=TaskStatus.OK)
+                return
+
+            self.redis.setex(
+                f"SubdomainEnumeration-done-{encoded_subdomain}", Config.Miscellaneous.SUBDOMAIN_ENUMERATION_TTL_DAYS * 24 * 60 * 60, 1
+            )
+            self.log.debug(f"Subdomain {subdomain} marked as done in Redis.")
+            print(f"Subdomain {subdomain} marked as done in Redis.")
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:129]: IMO this is a superfluous print

File: artemis/modules/subdomain_enumeration.py [hunk 38]
```diff
@@ -0,0 +1,145 @@
...
+
+        for subdomain in valid_subdomains:
+            task = Task(
+                {"type": TaskType.DOMAIN},
+                payload={
+                    "domain": subdomain,
+                },
+            )
+            self.log.debug(f"Creating task for subdomain: {subdomain} with payload: {task.payload}")
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:138]: why both log and print?

File: artemis/modules/subdomain_enumeration.py [hunk 39]
```diff
@@ -0,0 +1,145 @@
...
+            print(f"Creating task for subdomain: {subdomain} with payload: {task.payload}")
+            self.add_task(current_task, task)
+        
+        self.db.save_task_result(task=current_task, status=TaskStatus.OK, data=list(valid_subdomains))
+
+if __name__ == "__main__":
+    SubdomainEnumeration().loop()
```
Comment (kazet) [inline artemis/modules/subdomain_enumeration.py:145]: pls make sure linter and tests passes

**Comment by kazet** [comment]
btw remember to add --scale for subdomain enumeration in scripts/start so that it runs e.g. in 5 instances as it's slower

**Review by bulkowy — DISMISSED** [review]
